### PR TITLE
HW15 is completed

### PR DESCRIPTION
**Pull request by sterligov**
### Критерии оценки
- [x] Проект полностью запускается и останавливается с помощью `make up` / `make down` - 3 балла
- [x] Интеграционные тесты запускаются с помощью `make integration-tests`. Команда возвращает верный код ответа - 1 балл

Интеграционные тесты покрывают бизнес сценарии:
- [x] добавление события и обработка бизнес ошибок - 2 балла
- [x] получение листинга событий на день/неделю/месяц - 2 балла
- [x] отправка уведомлений - 2 балла

### CODE REVIEW

**Comment by sterligov** [comment]
Attaching to deployments_calendar_integration_1
calendar_integration_1  | === RUN   TestIntegration
calendar_integration_1  | === RUN   TestIntegration/TestCreateEvent
calendar_integration_1  | === RUN   TestIntegration/TestCreateEvent/ok
calendar_integration_1  | === RUN   TestIntegration/TestCreateEvent/busy_date_error
calendar_integration_1  | === RUN   TestIntegration/TestDeleteEventByID
calendar_integration_1  | === RUN   TestIntegration/TestDeleteEventByID/ok
calendar_integration_1  | === RUN   TestIntegration/TestDeleteEventByID/not_existing_event
calendar_integration_1  | === RUN   TestIntegration/TestGetEventByID
calendar_integration_1  | === RUN   TestIntegration/TestGetEventByID/ok
calendar_integration_1  | === RUN   TestIntegration/TestGetEventByID/not_existing_event
calendar_integration_1  | === RUN   TestIntegration/TestGetUserDayEvents
calendar_integration_1  | === RUN   TestIntegration/TestGetUserDayEvents/ok
calendar_integration_1  | === RUN   TestIntegration/TestGetUserDayEvents/no_events
calendar_integration_1  | === RUN   TestIntegration/TestGetUserMonthEvents
calendar_integration_1  | === RUN   TestIntegration/TestGetUserMonthEvents/ok
calendar_integration_1  | === RUN   TestIntegration/TestGetUserMonthEvents/no_events
calendar_integration_1  | === RUN   TestIntegration/TestGetUserWeekEvents
calendar_integration_1  | === RUN   TestIntegration/TestGetUserWeekEvents/ok
calendar_integration_1  | === RUN   TestIntegration/TestGetUserWeekEvents/no_events
calendar_integration_1  | === RUN   TestIntegration/TestSender
calendar_integration_1  | === RUN   TestIntegration/TestSender/notification
calendar_integration_1  | === RUN   TestIntegration/TestSender/delete_old_event
calendar_integration_1  | === RUN   TestIntegration/TestUpdateEvent
calendar_integration_1  | === RUN   TestIntegration/TestUpdateEvent/ok
calendar_integration_1  | === RUN   TestIntegration/TestUpdateEvent/not_existing_event
calendar_integration_1  | === RUN   TestIntegration/TestUpdateEvent/busy_date_error
calendar_integration_1  | --- PASS: TestIntegration (10.16s)
calendar_integration_1  |     --- PASS: TestIntegration/TestCreateEvent (0.04s)
calendar_integration_1  |         --- PASS: TestIntegration/TestCreateEvent/ok (0.03s)
calendar_integration_1  |         --- PASS: TestIntegration/TestCreateEvent/busy_date_error (0.01s)
calendar_integration_1  |     --- PASS: TestIntegration/TestDeleteEventByID (0.02s)
calendar_integration_1  |         --- PASS: TestIntegration/TestDeleteEventByID/ok (0.02s)
calendar_integration_1  |         --- PASS: TestIntegration/TestDeleteEventByID/not_existing_event (0.00s)
calendar_integration_1  |     --- PASS: TestIntegration/TestGetEventByID (0.01s)
calendar_integration_1  |         --- PASS: TestIntegration/TestGetEventByID/ok (0.00s)
calendar_integration_1  |         --- PASS: TestIntegration/TestGetEventByID/not_existing_event (0.00s)
calendar_integration_1  |     --- PASS: TestIntegration/TestGetUserDayEvents (0.01s)
calendar_integration_1  |         --- PASS: TestIntegration/TestGetUserDayEvents/ok (0.01s)
calendar_integration_1  |         --- PASS: TestIntegration/TestGetUserDayEvents/no_events (0.00s)
calendar_integration_1  |     --- PASS: TestIntegration/TestGetUserMonthEvents (0.01s)
calendar_integration_1  |         --- PASS: TestIntegration/TestGetUserMonthEvents/ok (0.00s)
calendar_integration_1  |         --- PASS: TestIntegration/TestGetUserMonthEvents/no_events (0.00s)
calendar_integration_1  |     --- PASS: TestIntegration/TestGetUserWeekEvents (0.01s)
calendar_integration_1  |         --- PASS: TestIntegration/TestGetUserWeekEvents/ok (0.00s)
calendar_integration_1  |         --- PASS: TestIntegration/TestGetUserWeekEvents/no_events (0.00s)
calendar_integration_1  |     --- PASS: TestIntegration/TestSender (10.00s)
calendar_integration_1  |         --- PASS: TestIntegration/TestSender/notification (0.00s)
calendar_integration_1  |         --- PASS: TestIntegration/TestSender/delete_old_event (0.00s)
calendar_integration_1  |     --- PASS: TestIntegration/TestUpdateEvent (0.03s)
calendar_integration_1  |         --- PASS: TestIntegration/TestUpdateEvent/ok (0.02s)
calendar_integration_1  |         --- PASS: TestIntegration/TestUpdateEvent/not_existing_event (0.00s)
calendar_integration_1  |         --- PASS: TestIntegration/TestUpdateEvent/busy_date_error (0.01s)
calendar_integration_1  | PASS
calendar_integration_1  | ok    github.com/sterligov/otus_homework/hw12_13_14_15_calendar/tests/integration     10.165s
deployments_calendar_integration_1 exited with code 0

File: hw12_13_14_15_calendar/Makefile [hunk 1]
```diff
@@ -82,16 +80,13 @@ wire:
...
-
 migrations:
-	goose -dir migrations mysql "${DB_USER}:${DB_PASSWORD}@tcp(${DB_HOST}:${DB_PORT})/${DB_NAME}?parseTime=true" up
+	goose -dir migrations mysql "${DB_USER}:${DB_PASSWORD}@tcp(localhost:${DB_PORT})/${DB_NAME}?parseTime=true" up
 
 generate:
 	go generate ./...
 
+integration-tests:
+	  chmod +x ./scripts/integration-test.sh && ./scripts/integration-test.sh
```
Comment (Antonboom) [inline hw12_13_14_15_calendar/Makefile:90]: вы можете один раз сделать chmod и закоммитить

флаги файла улетят в SVC

File: hw12_13_14_15_calendar/cmd/calendar/main.go [hunk 2]
```diff
@@ -56,27 +57,30 @@ func main() {
 	}
 	defer cleanup()
 
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals, syscall.SIGINT)
+
 	go func() {
 		if err := server.GRPC.Start(); err != nil {
 			logrus.Warnf("grpc server start failed: %s", err)
+			log.Fatalln(err)
```
Comment (Antonboom) [inline hw12_13_14_15_calendar/cmd/calendar/main.go:66]: при ошибке старта одного из серверов:
- не завершается аккуратно уже работающий сервис
- не завершаются аккуратно воркеры, подвязанные на рутовый контекст

File: hw12_13_14_15_calendar/internal/storage/memory/event.go [hunk 3]
```diff
@@ -67,6 +67,21 @@ func (es *EventStorage) UpdateEvent(_ context.Context, event storage.Event) (int
 	return 1, nil
 }
 
+func (es *EventStorage) UpdateIsNotified(_ context.Context, id storage.EventID, isNotified byte) error {
+	es.mu.Lock()
+	defer es.mu.Unlock()
+
+	for k, e := range es.bucket {
```
Comment (Antonboom) [inline hw12_13_14_15_calendar/internal/storage/memory/event.go:74]:
```
es.bucket[id]
```
?

File: hw12_13_14_15_calendar/tests/integration/integration_test.go [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+		testfixtures.Dialect("mysql"),
+		testfixtures.Directory("fixtures"),
+	)
+	s.Require().NoError(err)
+
+	err = fixtures.Load()
+	s.Require().NoError(err)
+}
+
+func (s *Suite) TearDownSuite() {
```
Comment (Antonboom) [inline hw12_13_14_15_calendar/tests/integration/integration_test.go:59]: не хватает контекста сьюта и его отмены

File: hw12_13_14_15_calendar/tests/integration/integration_test.go [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+	s.Require().NoError(s.db.Close())
+	s.Require().NoError(s.grpcConn.Close())
+}
+
+func (s *Suite) TestGetEventByID() {
+	s.Run("ok", func() {
+		resp, err := s.eventClient.GetEventByID(context.Background(), &pb.GetEventByIDRequest{
```
Comment (Antonboom) [inline hw12_13_14_15_calendar/tests/integration/integration_test.go:66]: вот здесь и в аналогичных местах использовать бы контекст сьюта

**Review by Antonboom — APPROVED** [review]
https://otus.ru/teacher-lk/homework/124257/8096/

Здравствуйте!
Хорошая работа.

Принято
10 / 10

File: hw12_13_14_15_calendar/Makefile (hunk 1 above)
Comment (sterligov) [reply hw12_13_14_15_calendar/Makefile:90]: Забыл про это)

File: hw12_13_14_15_calendar/internal/storage/memory/event.go (hunk 3 above)
Comment (sterligov) [reply hw12_13_14_15_calendar/internal/storage/memory/event.go:74]: Затупил)

**Pull request merged**